Add IteratorIndexable trait

fcard opened the pull request:
I was asked to make this preliminary PR as I required this functionality. This trait checks for the ability to index an iterator to obtain the values that could've been obtained from iteration. There are two subtypes:

* `NotIndexable` - For any iterator that cannot be indexed or its indexable value is not its eltype.
* `IsIndexable` - For iterators that define `getindex` such that `getindex(iterator, idxs...)::eltype(iterator)`, for some `idxs`.

I've added two other traits, as wew as two functions. The traits are:

* `EachIndexSupport`, to check that the type defines `eachindex`
* `AdjacentIndexSupport`, to check that the type defines one of `nextind`, `prevind`, or both.

The functions are:

* `eachindextype`, which is `typeof(eachindex(value))` for values that have `eachindex`, or missing otherwise.
* `eachindex_iteratortype`, which is `eachindextype(value)` for indexable iterators, or missing otherwise.

The reason for these added traits and functions are that just knowing that a iterator is indexable is not enough for many applications, for example when implementing a optimized drop, you'd need to know how get the `nth` index of a iterator, which could be accomplished by either `eachindex` or `nextind`. My use case, a reversable `Take` and `TakeWhile`, requires both.

Use cases include a [drop without overhead for certain indexable iterators,](https://gist.github.com/fcard/2756d9895ba450a4f1880812b4fa47b6#file-fastdrop-jl-L17-L54) or a [default reverse for all indexable iterators. (that have invertible eachindices)](https://gist.github.com/fcard/2756d9895ba450a4f1880812b4fa47b6#file-defaultreverse-jl-L18-L40) `IteratorIndexable` could also be used to [implement indexing in iterators like `Generator`*, `Take` and `Drop`.](https://gist.github.com/fcard/2756d9895ba450a4f1880812b4fa47b6#file-indexiter-jl)

`Indexable` is also not a word apparently, but it already exists in other languages serving the same purpose, like in swift, so I didn't put any more thought into it.

`*`Really more appropriately `Iterators.map`, but since they make up the same Type I couldn't separate them.

fcard committed:
Add IteratorIndexable trait

fcard commented:
I forgot a `()` in the doctests :(

fcard committed:
Fix IteratorIndexable doctest

dkarrasch commented on base/generator.jl:
```diff
@@ -128,3 +128,51 @@ IteratorEltype(::Type) = HasEltype()  # HasEltype is the default
+
+function IteratorIndexable(T::Type{<:AbstractArray{E,N}}) where {E,N}
+    if IndexStyle(T) === IndexLinear()
+        HasEachIndex{Base.OneTo{Int}}()
+    else
+        HasEachIndex{CartesianIndices{N, NTuple{N, Base.OneTo{Int}}}}()
```
Doesn't this assume one-based indexing? What about OffsetArrays?

fcard replied:
It does, I tried mirroring the default values for `eachindex` for each type but that's not really good enough, hmmm.. Maybe `AbstractArray` should default to `Indexable()` and then concrete Base types can use these default values.

dkarrasch replied:
I'm not sure, but can't you simply ask for `typeof(eachindex(A))`? Can't that be obtained at compile time?

fcard replied:
How about this?
```julia
struct UnknownEachIndex <: IteratorIndexable end
struct KnownEachIndex{I} <: IteratorIndexable end
HasEachIndex{I} = Union{UnknownEachIndex, KnownEachIndex{I}}
HasEachIndex() = UnknownEachIndex()
HasEachIndex{I}() where I = KnownEachIndex{I}()

function IteratorIndexable(x)
    I = IteratorIndexable(typeof(x))
    I isa UnknownEachIndex ? HasEachIndex{typeof(eachindex)}() : I
end

IteratorIndexable(::Type) = NotIndexable()
IteratorIndexable(::Type{<:Ref}) = Indexable()
IteratorIndexable(::Type{<:AbstractArray}) = HasEachIndex()
IteratorIndexable(::Type{<:AbstractString}) = HasEachIndex()
IteratorIndexable(::Type{<:Tuple}) = HasEachIndex{Base.OneTo{Int}}()
IteratorIndexable(::Type{<:Number}) = HasEachIndex{Base.OneTo{Int}}()
```

fcard replied:
Actually, there's a problem with this, in that `(UnknownEachIndex <: HasEachIndex{Int}) == true`. I suppose this works better?
```julia
abstract type HasEachIndex{I} end
struct UnknownEachIndex <: HasEachIndex{Any} end
struct KnownEachIndex{I} <: HasEachIndex{I} end
# [...]
```

fcard replied:
Or did you mean that I should remove the typevar and ask for the type outside the trait? I mostly did it as a shortcut because I could easily include it, but since it turned out to be more complicated than I thought, than it's not worth the hassle.

dkarrasch replied:
I meant returning `HasEachIndex{typeof(eachindex(A))}`. Then you wouldn't need the branches.

fcard replied:
Well, I don't have `A` at the point I return the first `HasEachIndex`, I could if I semi-duplicated the methods like this:
```julia
IteratorIndexable(x) = IteratorIndexable(typeof(x))
IteratorIndexable(x::AbstractArray) = HasEachIndex{typeof(eachindex(x))}()
IteratorIndexable(x::AbstractString) = HasEachIndex{typeof(eachindex(x))}()

IteratorIndexable(::Type) = NotIndexable()
IteratorIndexable(::Type{<:Ref}) = Indexable()
IteratorIndexable(::Type{<:AbstractArray}) = HasEachIndex()
IteratorIndexable(::Type{<:AbstractString}) = HasEachIndex()
IteratorIndexable(::Type{<:Tuple}) = HasEachIndex{Base.OneTo{Int}}()
IteratorIndexable(::Type{<:Number}) = HasEachIndex{Base.OneTo{Int}}()
```
But is that necessary? I think that's an easy branch for the compiler to solve:
```julia
julia> @code_typed IteratorIndexable([1,2,3,4])
CodeInfo(
1 ─     Base.arraysize(x, 1)::Int64
└──     return $(QuoteNode(KnownEachIndex{Base.OneTo{Int64}}()))
) => KnownEachIndex{Base.OneTo{Int64}}
```

dkarrasch replied:
OTOH, what you currently have is perhaps fine for a fallback method. Other array types should simply implement their own specialized trait methods.

fcard replied:
That's what I thought, now I am worried about breakages if e.g. we add an optimization in `Base` that depends on the default result, [like this one for `drop`](https://gist.github.com/fcard/2756d9895ba450a4f1880812b4fa47b6#file-fastdrop-jl-L34-L46), and then `drop` will be broken for any array that changes the type of `eachindex` to something else other than `AbstractRange` until they implement `IteratorIndexable`.

However, I am a bit dissatisfied with this new one too because it will make `IteratorIndexable(x) != IteratorIndexable(typeof(x))` for some types, which seems weird because it's `==` for all the other ones, although maybe `IteratorIndexable(x)` could be seen as a separate function that adds instance information to the trait. Or maybe we should properly separate it into another generic function so `IteratorIndexable(x) == IteratorIndexable(typeof(x))` but there is a `iteratorindexable(x)` (or something) function that expands the trait with instance information.

Hmmm..

fcard replied:
I've thought and decided that the type of `eachindex(itr)` doesn't belong in the trait for now. Instead I've added a sister function `eachindextype`:
```julia
eachindextype(x) = eachindextype(x, IteratorIndexable(x))
eachindextype(x, ::HasEachIndex) = typeof(eachindex(x))
eachindextype(x, ::IteratorIndexable) = missing
```
It's not a trait, since traits are meant to be based on the type of the argument (as far as I know), and the concrete return type of a function is a runtime property of instances. (It's why we can't just use `Base.return_types`, other than it being a private part of the inference engine)

fcard replied:
New issue: `eachindextype` would be only be defined for `x` if `IteratorIndexable(x) == HasEachIndex()`, but that means that `eachindextype(Dict(1=>2)) == missing`, despite there existing `eachindex(Dict(1=>2))`. That's intentional but could be confusing. Maybe we need a new name? `iterator_eachindex_type`? `eachindex_iteratortype`?

Maybe we need another trait, `EachIndexable` or something, that includes `Dict`, then `eachindextype` will use that trait instead, and then we combine constraints with `IteratorIndexable` when required.

```julia
iterate(itr::Drop, state...) = _iterate_drop(itr, state..., IteratorIndexable(itr), eachindextype(itr))

function _iterate_drop(it, ::IsIndexable, ::Type{<:AbstractRange})
    [...] # optimized drop
end

function _iterate_drop(it, _, _)
    [...] # default drop
end
```

fcard committed:
Add EachIndexSupport & AdjacentIndexSupport traits

fcard commented:
I've decided to add two other traits, `EachIndexSupport` for `eachindex` and `AdjacentIndexSupport` for both `prevind` and `nextind`. I've added a `eachindextype` function as wew as a `eachindex_iteratortype` function. That covers all my needs, so I have removed the WIP. Not sure on the names, specially `eachindex_iteratortype`, but I don't think I can come up with anything I will be confident with.

fcard commented:
Birthday bump! any thoughts on the new code? Is this acceptable or should I try merging them into a single trait again? I don't think I can make it as sensible as it is with three traits, but if that's too much I can try something.

DilumAluthge commented:
This PR has had no activity for four years+, so I'll close it as stale.

@fcard If you're still interested in working on this PR, let me know, and I can re-open it.
